File: user-interfaces/sdk/typescript/s3/examples/basic-usage.ts
/**
 * S3 SDK - Basic Usage Example
 *
 * This example demonstrates:
 * 1. Connecting to the blockchain and provider
 * 2. Creating a bucket
 * 3. Uploading objects
 * 4. Listing and downloading objects
 *
 * Prerequisites:
 *   - Parachain running at ws://127.0.0.1:2222
 *   - Provider running at http://127.0.0.1:3333
 *   - npm install && npm run papi:generate
 *
 * Usage:
 *   npx tsx examples/basic-usage.ts
 */

import { S3Client } from "../src/index.js";

const CHAIN_WS = process.env.CHAIN_WS || "ws://127.0.0.1:2222";
const PROVIDER_URL = process.env.PROVIDER_URL || "http://127.0.0.1:3333";

async function main() {
  console.log("=== S3 SDK - Basic Usage ===\n");
  console.log(`Chain: ${CHAIN_WS}`);
  console.log(`Provider: ${PROVIDER_URL}\n`);

  // Create client
  const client = new S3Client({
    chainWs: CHAIN_WS,
    providerUrl: PROVIDER_URL,
  });

  try {
    // Step 1: Connect
    console.log("Step 1: Connecting...");
    await client.connect();
    await client.setSigner("//Alice");
    console.log(`  Connected as: ${client.getAddress()}\n`);

    // Step 2: Create a bucket
    const bucketName = `test-bucket-${Date.now()}`;
    console.log(`Step 2: Creating bucket '${bucketName}'...`);
    const bucket = await client.createBucket(bucketName);
    console.log(`  Bucket created:`);
    console.log(`    S3 Bucket ID: ${bucket.s3BucketId}`);
    console.log(`    Layer 0 Bucket ID: ${bucket.layer0BucketId}`);
    console.log(`    Owner: ${bucket.owner}\n`);

    // Step 3: Upload objects
    console.log("Step 3: Uploading objects...");

    const textContent = new TextEncoder().encode("Hello from S3 SDK!");
    const result1 = await client.putObject(bucketName, "hello.txt", textContent, {
      contentType: "text/plain",
    });
    console.log(`  Uploaded 'hello.txt':`);
    console.log(`    CID: ${result1.cid}`);
    console.log(`    ETag: ${result1.etag}`);
    console.log(`    Size: ${result1.size} bytes`);

    const jsonContent = new TextEncoder().encode(
      JSON.stringify({ message: "Hello JSON!", timestamp: Date.now() })
    );
    const result2 = await client.putObject(
      bucketName,
      "data/config.json",
      jsonContent,
      {
        contentType: "application/json",
        metadata: { "x-custom-key": "custom-value" },
      }
    );
    console.log(`  Uploaded 'data/config.json':`);
    console.log(`    CID: ${result2.cid}`);
    console.log(`    Size: ${result2.size} bytes\n`);

    // Step 4: Get object metadata
    console.log("Step 4: Getting object metadata...");
    const metadata = await client.headObject(bucketName, "hello.txt");
    console.log(`  Object 'hello.txt' metadata:`);
    console.log(`    Size: ${metadata.size}`);
    console.log(`    Content-Type: ${metadata.contentType}`);
    console.log(`    ETag: ${metadata.etag}\n`);

    // Step 5: Download and verify
    console.log("Step 5: Downloading and verifying...");
    const downloaded = await client.getObject(bucketName, "hello.txt");
    const downloadedText = new TextDecoder().decode(downloaded.data);
    console.log(`  Downloaded: "${downloadedText}"`);

    const matches = downloadedText === "Hello from S3 SDK!";
    console.log(`  Verified: ${matches ? "OK" : "MISMATCH"}\n`);

    // Step 6: List buckets
    console.log("Step 6: Listing all buckets...");
    const buckets = await client.listBuckets();
    console.log(`  Found ${buckets.length} bucket(s):`);
    for (const b of buckets) {
      console.log(`    - ${b.name} (ID: ${b.s3BucketId}, Objects: ${b.objectCount})`);
    }

    console.log("\n=== Example completed successfully! ===");
  } catch (error) {
    console.error("\nError:", error);
    process.exitCode = 1;
  } finally {
    client.disconnect();
  }
}

main();

File: storage-interfaces/s3/client/examples/basic_usage.rs
//! Basic S3 Client Usage Example
//!
//! Usage: cargo run --example basic_usage [chain_ws] [provider_url] [seed]

use s3_client::{PutObjectOptions, S3Client};
use sp_runtime::AccountId32;
use std::collections::HashMap;
use std::env;
use storage_client::{NegotiateRequest, ProviderClient};
use subxt_signer::sr25519::dev as dev_signer;

const DEFAULT_CHAIN_WS: &str = "ws://127.0.0.1:2222";
const DEFAULT_PROVIDER_URL: &str = "http://127.0.0.1:3333";
const DEFAULT_SEED: &str = "//Alice";

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    tracing_subscriber::fmt::init();

    let args: Vec<String> = env::args().collect();
    let chain_url = args.get(1).map(|s| s.as_str()).unwrap_or(DEFAULT_CHAIN_WS);
    let provider_url = args
        .get(2)
        .map(|s| s.as_str())
        .unwrap_or(DEFAULT_PROVIDER_URL);
    let seed = args.get(3).map(|s| s.as_str()).unwrap_or(DEFAULT_SEED);

    println!("=== S3 Client Basic Usage Example ===\n");
    println!("Chain URL: {chain_url}");
    println!("Provider URL: {provider_url}");
    println!("Account: {seed}\n");

    println!("Creating S3 client...");
    let client = S3Client::new(chain_url, provider_url, seed).await?;
    println!("S3 client created successfully!\n");

    // Resolve owner/provider accounts. Owner derives from `seed`; provider is
    // the dev account the local node was started with (//Bob by default).
    let owner: AccountId32 = match seed {
        "//Alice" => dev_signer::alice().public_key().0.into(),
        "//Bob" => dev_signer::bob().public_key().0.into(),
        "//Charlie" => dev_signer::charlie().public_key().0.into(),
        "//Dave" => dev_signer::dave().public_key().0.into(),
        "//Eve" => dev_signer::eve().public_key().0.into(),
        "//Ferdie" => dev_signer::ferdie().public_key().0.into(),
        _ => return Err(format!("unsupported seed for this example: {seed}").into()),
    };
    // Discover the provider's on-chain account from its /info endpoint
    let provider = ProviderClient::fetch_provider_id(provider_url).await?;
    println!("  Provider account (from /info): {provider}");

    println!("Negotiating signed agreement terms with provider...");
    let signed = ProviderClient::negotiate_terms(
        provider_url,
        &NegotiateRequest {
            owner,
            max_bytes: 1_000_000_000, // 1 GB
            duration: 500,            // 500 blocks
            price_per_byte: 0,
            replica_params: None,
            bucket_id: None,
        },
    )
    .await?;
    println!(
        "Provider signed terms: nonce={}, valid_until={}\n",
        signed.terms.nonce, signed.terms.valid_until
    );

    let bucket_name = format!(
        "test-bucket-{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)?
            .as_secs()
    );

    println!("Creating bucket: {bucket_name}");
    let bucket = client
        .create_bucket(&bucket_name, provider, signed.terms, signed.signature)
        .await?;
    println!("Bucket created:");
    println!("  S3 Bucket ID: {}", bucket.s3_bucket_id);
    println!("  Layer 0 Bucket ID: {}", bucket.layer0_bucket_id);
    println!();

    println!("Uploading object: hello.txt");
    let content = b"Hello, Web3 Storage!";
    let mut metadata = HashMap::new();
    metadata.insert("x-custom-key".to_string(), "custom-value".to_string());

    let put_result = client
        .put_object(
            &bucket_name,
            "hello.txt",
            content,
            PutObjectOptions {
                content_type: Some("text/plain".to_string()),
                metadata,
            },
        )
        .await?;

    println!("Object uploaded:");
    println!("  ETag: {}", put_result.etag);
    println!("  CID: {:?}", put_result.cid);
    println!("  Size: {} bytes", put_result.size);
    println!();

    println!("Downloading object: hello.txt");
    let get_result = client.get_object(&bucket_name, "hello.txt").await?;
    println!("Object downloaded:");
    println!("  Content: {}", String::from_utf8_lossy(&get_result.data));
    println!("  Size: {} bytes", get_result.size);
    println!();

    println!("Cleaning up...");
    client.delete_object(&bucket_name, "hello.txt").await?;
    println!("Object deleted");

    client.delete_bucket(&bucket_name).await?;
    println!("Bucket deleted");
    println!();

    println!("=== Example completed successfully! ===");
    Ok(())
}
